cumulusci/tasks/push/tests/test_push_api.py

```diff
@@ -1,12 +1,541 @@
+    )
+    assert mock.call(batch_0, push_request_id) in sf_push_api._add_batch.call_args_list
+    assert mock.call(batch_1, push_request_id) in sf_push_api._add_batch.call_args_list
+    assert push_request_id == actual_id
+    assert 2 == actual_org_count
+
+
+def test_sf_push_add_push_batch(sf_push_api, metadata_package_version):
+    push_request_id = "0DV?xxxxxx?"
+    metadata_package_version.sf_id = "0KM?xxxxx?"
+    orgs = ["00D000000001", "00D000000002"]
+    expected_records_json = json.dumps(
+        {
+            "records": [
+                {
+                    "attributes": {"type": "PackagePushJob", "referenceId": orgs[0]},
+                    "PackagePushRequestId": push_request_id,
+                    "SubscriberOrganizationKey": orgs[0],
+                },
+                {
+                    "attributes": {"type": "PackagePushJob", "referenceId": orgs[1]},
+                    "PackagePushRequestId": push_request_id,
+                    "SubscriberOrganizationKey": orgs[1],
+                },
+            ]
+        }
+    )
+
+    sf_push_api.sf.base_url = "base_url/"
+
+    returned_batch = sf_push_api._add_batch(orgs, push_request_id)
+    sf_push_api.sf._call_salesforce.assert_called_once_with(
+        "POST", "base_url/composite/tree/PackagePushJob", data=expected_records_json
+    )
+    assert ["00D000000001", "00D000000002"] == returned_batch
+
+
+def test_sf_push_add_push_batch_retry(sf_push_api, metadata_package_version):
+    push_request_id = "0DV?xxxxxx?"
+    orgs = ["00D000000001", "00D000000002", "00D000000003"]
+    retry_response = {
+        "results": [
+            {
+                "referenceId": orgs[0],
+                "errors": [
+                    {"message": "Something bad has happened! Whatever could it be?"}
+                ],
+            }
+        ]
+    }
+    duplicate_response = {
+        "results": [
+            {
+                "referenceId": orgs[1],
+                "errors": [{"message": "", "statusCode": "DUPLICATE_VALUE"}],
+            }
+        ]
+    }
+    invalid_response = {
+        "results": [
+            {
+                "referenceId": orgs[2],
+                "errors": [{"message": "", "statusCode": "INVALID_OPERATION"}],
+            }
+        ]
+    }
+
+    sf_push_api.sf.base_url = "base_url/"
+    sf_push_api.sf._call_salesforce.side_effect = [
+        SalesforceMalformedRequest(
+            "base_url/composite/tree/PackagePushJob",
+            401,
+            "resource_name",
+            retry_response,
+        ),
+        SalesforceMalformedRequest(
+            "base_url/composite/tree/PackagePushJob",
+            401,
+            "resource_name",
+            duplicate_response,
+        ),
+        SalesforceMalformedRequest(
+            "base_url/composite/tree/PackagePushJob",
+            401,
+            "resource_name",
+            invalid_response,
+        ),
+        [],
+    ]
```

Comment:
401 is an Unauthorized response, which is not what the API would be returning for these errors.

Reply:
I think these should be 400, since that's the status code that simple-salesforce transforms into a SalesforceMalformedRequest

Reply:
@davisagli where did you look to figure this out?

Reply:
simple-salesforce source code: https://github.com/simple-salesforce/simple-salesforce/blob/master/simple_salesforce/util.py#L51